lib/api/model/submessage.dart

```diff
@@ -0,0 +1,285 @@
+  poll,
+  unknown;
+
+  factory WidgetType.fromJson(Object value) {
+    return $enumDecode<WidgetType, String>(
+      _$WidgetTypeEnumMap,
+      value,
+      unknownValue: WidgetType.unknown
+    );
+  }
+}
+
+/// The data encoded in a submessage to make the message a poll widget.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollWidgetData extends WidgetData {
+  @override
+  @JsonKey(includeToJson: true)
+  WidgetType get widgetType => WidgetType.poll;
+
+  /// The initial question and options on the poll.
+  final PollWidgetExtraData extraData;
+
+  PollWidgetData({required this.extraData});
+
+  factory PollWidgetData.fromJson(Map<String, Object?> json) =>
+    _$PollWidgetDataFromJson(json);
+
+  @override
+  Map<String, Object?> toJson() => _$PollWidgetDataToJson(this);
+}
+
+/// As in [PollWidgetData.extraData].
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollWidgetExtraData {
+  final String question;
+  final List<String> options;
+
+  const PollWidgetExtraData({required this.question, required this.options});
+
+  factory PollWidgetExtraData.fromJson(Map<String, Object?> json) =>
+    _$PollWidgetExtraDataFromJson(json);
+
+  Map<String, Object?> toJson() => _$PollWidgetExtraDataToJson(this);
+}
+
+class UnsupportedWidgetData extends WidgetData {
+  @override
+  @JsonKey(includeToJson: true)
+  WidgetType get widgetType => WidgetType.unknown;
+
+  UnsupportedWidgetData({required this.json});
+
+  final Object? json;
+
+  @override
+  Object? toJson() => json;
+}
+
+/// The data encoded in a submessage that acts on a poll.
+sealed class PollEvent extends SubmessageData {
+  PollEventType get type;
+
+  PollEvent();
+
+  /// The key for identifying the [optionIndex]'th option added by user
+  /// [senderId] to a poll.
+  ///
+  /// For options that are a part of the initial [PollWidgetData], the
+  /// [senderId] should be `null`.
+  static String optionKey({required int? senderId, required int optionIndex}) =>
+    // "canned" is a canonical constant coined by the web client.
+    '${senderId ?? 'canned'},$optionIndex';
+
+  factory PollEvent.fromJson(Map<String, Object?> json) {
+    final rawPollEventType = json['type'];
+    if (rawPollEventType == null) return UnknownPollEvent(json: json);
+    switch ($enumDecode(
+      _$PollEventTypeEnumMap,
+      rawPollEventType,
+      unknownValue: PollEventType.unknown,
+    )) {
+      case PollEventType.newOption: return PollOptionEvent.fromJson(json);
+      case PollEventType.question: return PollQuestionEvent.fromJson(json);
+      case PollEventType.vote: return PollVoteEvent.fromJson(json);
+      case PollEventType.unknown: return UnknownPollEvent(json: json);
+    }
+  }
+
+  Map<String, Object?> toJson();
+}
+
+/// A poll event when an option is added.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollOptionEvent extends PollEvent {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventType get type => PollEventType.newOption;
```

Comment:
If the `type` field says "new option", let's say "new option" in our class name too. (Just like we do with `Event` subclasses.) That helps with correlating our logic with the API docs when they exist, or with other clients when those don't.

And in this case it makes for an intrinsically better name, too — if the semantics are (as the doc comment says) that this is *adding* an option, and not doing something else like removing or editing one, then it's good for the name to reflect that.

So:
```suggestion
class PollNewOptionEvent extends PollEvent {
  @override
  @JsonKey(includeToJson: true)
  PollEventType get type => PollEventType.newOption;
```

(modulo the other comment about the endings of these names)